source/hp/dof_handler.cc

```diff
@@ -1745,55 +1724,160 @@ namespace hp
+                // Store the active_fe_index of each cell that will be refined
+                // to and distribute it later on its children.
+                refined_cells_fe_index.insert({cell, cell->active_fe_index()});
+              }
+            else if (cell->coarsen_flag_set())
+              {
+                // From all cells that will be coarsened, determine their parent
+                // and calculate its proper active_fe_index, so that it can be
+                // set after refinement.
+                // But first, check if that particular cell has a parent at all.
+                Assert(cell->level() > 0, ExcInternalError());
+                const auto &parent = cell->parent();
+                // Check if the active_fe_index for the current cell has been
+                // determined already.
+                if (coarsened_cells_fe_index.find(parent) ==
+                    coarsened_cells_fe_index.end())
+                  {
+                    std::set<unsigned int> fe_indices_children;
+                    for (unsigned int child_index = 0;
+                         child_index < parent->n_children();
+                         ++child_index)
+                      fe_indices_children.insert(
+                        parent->child(child_index)->active_fe_index());
+
+                    const unsigned int fe_index =
+                      fe_collection.find_least_dominating_fe_in_collection(
+                        fe_indices_children, /*codim=*/0);
+
+                    Assert(
+                      fe_index != numbers::invalid_unsigned_int,
+                      ExcMessage(
+                        "No FiniteElement has been found in your FECollection "
+                        "that dominates all children of a cell you are trying "
+                        "to coarsen!"));
+
+                    coarsened_cells_fe_index.insert({parent, fe_index});
+                  }
               }
           }
+
+        // Finally, restore current subdomain_ids.
+        if (shared_tria != nullptr && shared_tria->with_artificial_cells())
+          for (auto &cell : active_cell_iterators())
+            {
+              if (cell->is_artificial())
+                cell->set_subdomain_id(numbers::invalid_subdomain_id);
+              else
+                cell->set_subdomain_id(
+                  saved_subdomain_ids[cell->active_cell_index()]);
+            }
       }
+  }
+
+
+  template <int dim, int spacedim>
+  void
+  DoFHandler<dim, spacedim>::post_refinement_fe_index_update()
+  {
+    // Finite elements need to be assigned to each cell by calling
+    // distribute_dofs() first to make this functionality available.
+    if (fe_collection.size() > 0)
+      {
+        // Distribute active_fe_indices from all refined cells on their
+        // respective children.
+        for (const auto &pair : refined_cells_fe_index)
+          {
+            const cell_iterator parent(*(pair.first), this);
```

Comment:
I *think* you can get away by just writing
```
  const cell_iterator parent = pair.first;
```
I'm not going to hold up the whole patch about this, but would appreciate if you could try this out as a follow-up.